File: SortingAlgorithm/_13_PigeonholeSort/PigeonholeSort.py
def pigeonhole_sort(a):
    """
    Pigeonhole Sort implementation.
    
    Args:
    a (list): The input list to be sorted.

    Note:
    - This function sorts the input list in place.
    - It assumes that the input list contains only integers.
    """
    # Find the range of values in the list
    min_ele = min(a)
    max_ele = max(a)
    size = max_ele - min_ele + 1

    # Create pigeonholes
    holes = [0] * size

    # Populate the pigeonholes
    for x in a:
        assert isinstance(x, int), "Integers only, please"
        holes[x - min_ele] += 1

    # Put the elements back into the array in order
    i = 0
    for count in range(size):
        while holes[count] > 0:
            holes[count] -= 1
            a[i] = count + min_ele
            i += 1


# Test the sorting algorithm
a = [8, 3, 2, 7, 4, 6, 8]
print("Sorted order is:", end=' ')
pigeonhole_sort(a)
print(*a, end=' ')

File: SortingAlgorithm/_13_PigeonholeSort/PigeonholeSort.java
package Algorithm.SortingAlgorithm._13_PigeonholeSort;

import java.util.Arrays;

public class PigeonholeSort {

    public static void pigeonholeSort(int arr[]) {
        int min = arr[0];
        int max = arr[0];

        for (int a = 0; a < arr.length; a++) {
            if (arr[a] > max) {
                max = arr[a];
            }
            if (arr[a] < min) {
                min = arr[a];
            }
        }

        int range = max - min + 1;
        int[] pigeonholes = new int[range];
        Arrays.fill(pigeonholes, 0);

        for (int i = 0; i < arr.length; i++) {
            pigeonholes[arr[i] - min]++;
        }

        int index = 0;

        for (int j = 0; j < range; j++) {
            while (pigeonholes[j]-- > 0) {
                arr[index++] = j + min;
            }
        }
    }

    public static void main(String[] args) {
        PigeonholeSort sort = new PigeonholeSort();
        int[] arr = {8, 3, 2, 7, 4, 6, 8};

        System.out.print("Sorted order is: ");
        sort.pigeonholeSort(arr);

        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
    }
}
